### PR TITLE
Curator 161

### PR DESCRIPTION
**Pull request by cammckenzie**
First cut of watcher removal. Still need to provide the ability to register for notification of watch removal.

### CODE REVIEW

File: curator-framework/src/main/java/org/apache/curator/framework/CuratorFramework.java [hunk 1]
```diff
@@ -1,233 +1,238 @@
-/**
```
Comment (Randgalt) [inline curator-framework/src/main/java/org/apache/curator/framework/CuratorFramework.java]: Any idea why this looks like a complete replacement?

Comment (cammckenzie) [inline curator-framework/src/main/java/org/apache/curator/framework/CuratorFramework.java]: My only guess is that it's some sort of whitespace issue?

File: curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesBuilder.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+     * @param watcher
+     * @return
+     */
+    public RemoveWatchesType watcher(CuratorWatcher watcher);
+    
+    /**
+     * Specify that all watches should be removed
+     * @return
+     */
+    public RemoveWatchesType allWatches();
```
Comment (Randgalt) [inline curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesBuilder.java]: I'd rename this to something like "all". Try to make these build sentences: client.removeWatches().all()

File: curator-framework/src/main/java/org/apache/curator/framework/CuratorFramework.java [hunk 3]
```diff
@@ -1,233 +1,238 @@
...
+     *
+     * @return builder object
+     */
+    public SyncBuilder sync();
+
+    /**
+     * Start a remove watches builder.
+     * @return builder object
+     */
+    public RemoveWatchesBuilder removeWatches();
```
Comment (Randgalt) [inline curator-framework/src/main/java/org/apache/curator/framework/CuratorFramework.java]: Rename this to something like "watches" so we can get nice sentences: client.watches().removeAll()

File: curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesBuilder.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+ * Builder to allow watches to be removed 
+ */
+public interface RemoveWatchesBuilder
+{
+    /**
+     * Specify the watcher to be removed
+     * @param watcher
+     * @return
+     */
+    public RemoveWatchesType watcher(Watcher watcher);
```
Comment (Randgalt) [inline curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesBuilder.java]: I'd rename this to something like "remove". Try to make these build sentences: client.watches().remove(w).

File: curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesBuilder.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    
+    /**
+     * Specify the watcher to be removed
+     * @param watcher
+     * @return
+     */
+    public RemoveWatchesType watcher(CuratorWatcher watcher);
```
Comment (Randgalt) [inline curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesBuilder.java]: I'd rename this to something like "remove". Try to make these build sentences: client.watches().remove(w).

**Comment by Randgalt** [comment]
I think this change should also use the remove watches APIs in the various recipes. Or should that be a separate issue?

**Comment by cammckenzie** [comment]
Thanks for looking at this. I think that maybe a new JIRA for the recipes.
(I think there's already one for the node cache recipes).
cheers

**Comment by Randgalt** [comment]
OK - please open an issue for this. If the Node Cache issue is appropriate, rename it and target it for 3.0.0

File: curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesLocal.java [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+public interface RemoveWatchesLocal extends BackgroundPathable<Void>
+{
+   
+    /**
+     * Specify if the client should just remove client side watches if a connection to ZK
+     * is not available.
+     * @param local
+     * @return
+     */
+    public BackgroundPathable<Void> local(boolean local);
```
Comment (Randgalt) [inline curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesLocal.java]: the boolean doesn't seem necessary. If local() isn't called then it's not local.

Comment (Randgalt) [inline curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesLocal.java]: I'd name this "locally" so that it reads right. E.g. {{client.watches().remove(w).ofType(t).locally().forPath(...)}}

**Comment by cammckenzie** [comment]
I have made the changes based on your suggestions.

Was wondering if we wanted to have some sort of way of registering for notification of a watch being removed, or if we just do it via the CuratorListener?

Currently you need to register a CuratorListener and then check

if(event.getType() == CuratorEventType.WATCHED && event.getWatchedVent().getType() == EventType.DataWatchRemoved)
{
}

(or EventType.ChildWatchRemoved).

Maybe a new type of listener interface? Or is this overkill?

**Comment by cammckenzie** [comment]
There's actually another interesting thing that I hadn't thought about. The 'local' flag lets a client remove watches while its not actually connected to ZK, but the current way that callWithRetry() works is that it blocks until the client is connected or timed out prior to executing the Callable.

I guess I need to add something to Callable to indicate that it should be called even if there's no connection to ZK available. Does that sound reasonable?

**Comment by Randgalt** [comment]
You can do this to get the ZooKeeper handle directly:

```
client.getZookeeperClient().getZooKeeper()
```

This does not use the retry mechanism. So, when you are in "local" mode, don't do a retry. Just get the ZK handle and make the call. IMPORTANT: getZooKeeper() can legally return <code>null</code> in some cases.

File: curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesBuilder.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
+package org.apache.curator.framework.api;
```
Comment (Randgalt) [inline curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesBuilder.java]: This will need a guaranteed() method similar to node deletion. Use the FailedDeleteManager as a model (maybe extend it?).

File: curator-framework/src/main/java/org/apache/curator/framework/imps/RemoveWatchesBuilderImpl.java [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+    private void pathInForeground(final String path) throws Exception
+    {
+        RetryLoop.callWithRetry(client.getZookeeperClient(), 
+                new Callable<Void>()
+                {
+                    @Override
+                    public Void call() throws Exception
+                    {
+                        ZooKeeper zkClient = client.getZooKeeper();
+                        if(watcher == null)
```
Comment (Randgalt) [inline curator-framework/src/main/java/org/apache/curator/framework/imps/RemoveWatchesBuilderImpl.java]: Please review. if watcher is null, removeAllWatches is done. If not null, removeWatches.

File: curator-framework/src/main/java/org/apache/curator/framework/imps/RemoveWatchesBuilderImpl.java [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+                trace.commit();
+                CuratorEvent event = new CuratorEventImpl(client, CuratorEventType.REMOVE_WATCHES, rc, path, null, ctx, null, null, null, null, null);
+                client.processBackgroundOperation(operationAndData, event);                
+            }
+        };
+        
+        ZooKeeper zkClient = client.getZooKeeper();
+        if(watcher == null)
+        {
+            zkClient.removeAllWatches(operationAndData.getData(), watcherType, local, callback, operationAndData.getContext());
```
Comment (Randgalt) [inline curator-framework/src/main/java/org/apache/curator/framework/imps/RemoveWatchesBuilderImpl.java]: Please review. if watcher is null, removeAllWatches is done. If not null, removeWatches.

File: curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesBuilder.java (hunk 7 above)
Comment (Randgalt) [inline curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesBuilder.java]: We should also add a <code>quietly()</code> method or something to swallow <code>KeeperException.NoWatcher</code> exceptions. I'll bet most users don't care if the watcher doesn't exist.

File: curator-framework/src/main/java/org/apache/curator/framework/imps/RemoveWatchesBuilderImpl.java (hunk 8 above)
Comment (cammckenzie) [inline curator-framework/src/main/java/org/apache/curator/framework/imps/RemoveWatchesBuilderImpl.java]: This is reasonable yes? If we are specifying a particular watcher to remove then do that, otherwise remove all watches from the path.

File: curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesBuilder.java (hunk 7 above)
Comment (cammckenzie) [inline curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesBuilder.java]: I think quietly() is a good idea as its explicit. Also, guaranteed() will never tell the user that the remove has failed, whereas quietly would indicate failure (except for NoWatcher exceptions).

Comment (cammckenzie) [inline curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesBuilder.java]: Under what conditions can getZookeeperClient().getZookeeper() return null? Is it when we're trying to reconnect?

What should be done under these conditions? Presumably we just need to throw an exception back to the user because  we can't execute the command.

File: curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesLocal.java [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+ */
+public interface RemoveWatchesLocal extends BackgroundPathableQuietly<Void>
+{
+   
+    /**
+     * Specify if the client should just remove client side watches if a connection to ZK
+     * is not available.
+     * @return
+     */
+    public BackgroundPathableQuietly<Void> local();
```
Comment (Randgalt) [inline curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesLocal.java]: In context, this would be better named "locally". E.g.:

```
client.watches().removeAll().ofType(WatcherType.Any).locally().forPath(path);
```

File: curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesType.java [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+ */
+public interface RemoveWatchesType
+{
+   
+    /**
+     * Specify the type of watcher to be removed.
+     * @param watcherType
+     * @return
+     */
+    public RemoveWatchesLocal ofType(WatcherType watcherType);
```
Comment (Randgalt) [inline curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesType.java]: Can this be made optional with a default of WatcherType.Any?

File: curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesLocal.java (hunk 10 above)
Comment (cammckenzie) [inline curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesLocal.java]: Done

File: curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesType.java (hunk 11 above)
Comment (cammckenzie) [inline curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesType.java]: Done, it will now accept:
client.watches().removeAll().forPath("/");

File: curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesBuilder.java (hunk 7 above)
Comment (cammckenzie) [inline curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesBuilder.java]: Looking at the local case again, I think it's actually just client.getZookeeper() that needs to be used. I notice that this is how it's used in all of the performBackgroundOperation calls, but there are no checks for null pointers in this case. Is there a reason for this?

Comment (Randgalt) [inline curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesBuilder.java]: Maybe I mis-remembered. If the rest of the code doesn't check you don't need to.

File: curator-framework/src/main/java/org/apache/curator/framework/imps/RemoveWatchesBuilderImpl.java [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+        RetryLoop.callWithRetry(client.getZookeeperClient(), 
+                new Callable<Void>()
+                {
+                    @Override
+                    public Void call() throws Exception
+                    {
+                        try
+                        {
+                            ZooKeeper zkClient = client.getZooKeeper();
+                            if(watcher == null)
```
Comment (Randgalt) [inline curator-framework/src/main/java/org/apache/curator/framework/imps/RemoveWatchesBuilderImpl.java]: Don't forget to look at these lines. It seems wrong to me.

File: curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesBuilder.java (hunk 7 above)
Comment (cammckenzie) [inline curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesBuilder.java]: There are additional problems when trying the local case in the background as the 'performBackgroundOperation()' method checks if the client is connected prior to executing the operation.

File: curator-framework/src/main/java/org/apache/curator/framework/imps/RemoveWatchesBuilderImpl.java [hunk 13]
```diff
@@ -0,0 +1,214 @@
+package org.apache.curator.framework.imps;
```
Comment (Randgalt) [inline curator-framework/src/main/java/org/apache/curator/framework/imps/RemoveWatchesBuilderImpl.java]: Once the builder is committed via forPath() the watcher should also be removed from NamespaceWatcherMap. We should deprecate clearWatcherReferences() too.

File: curator-framework/src/main/java/org/apache/curator/framework/imps/RemoveWatchesBuilderImpl.java [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+        this.watcherType = null;
+        this.local = false;
+        this.quietly = false;
+        this.backgrounding = new Backgrounding();
+    }
+    
+    @Override
+    public RemoveWatchesType remove(Watcher watcher)
+    {
+        this.watcher = watcher == null ? null : client.getNamespaceWatcherMap().getNamespaceWatcher(watcher);
```
Comment (Randgalt) [inline curator-framework/src/main/java/org/apache/curator/framework/imps/RemoveWatchesBuilderImpl.java]: <code> client.getNamespaceWatcherMap().getNamespaceWatcher(watcher)</code> will add a namespace watcher record if it doesn't exist. That doesn't seem right for removal. A new method in NamespaceWatcherMap is probably needed.

Comment (cammckenzie) [inline curator-framework/src/main/java/org/apache/curator/framework/imps/RemoveWatchesBuilderImpl.java]: I guess if we don't have a reference in the NamespaceWatcherMap then we pass the watcher through to the ZK client without it being wrapper in a NamespaceWatcher. The only reason I can think of a watcher not being in the map would be if it was registered directly with the ZK client without going through the ZK framework. Either way, if it doesn't exist it will result in a NoWatcherException

File: curator-framework/src/main/java/org/apache/curator/framework/imps/RemoveWatchesBuilderImpl.java (hunk 13 above)
Comment (cammckenzie) [inline curator-framework/src/main/java/org/apache/curator/framework/imps/RemoveWatchesBuilderImpl.java]: I don't think that this is necessarily true is it? We don't want to remove the Watcher reference itself because it may still be watching other paths. If it is no longer referenced then the WeakReference in the NamespaceWatcherMaps internal cache should handle this.

File: curator-framework/src/main/java/org/apache/curator/framework/imps/RemoveWatchesBuilderImpl.java (hunk 12 above)
Comment (cammckenzie) [inline curator-framework/src/main/java/org/apache/curator/framework/imps/RemoveWatchesBuilderImpl.java]: What is it that seems wrong? If no Watcher is specified this means that we're removing all watches for a given path. If a Watcher is specified then this means we're removing a specific watch from a given path.

File: curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesBuilder.java (hunk 7 above)
Comment (Randgalt) [inline curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesBuilder.java]: Internally, it looks like "local" mode acts only if the connection is lost. Maybe you can test the connection state when local is set and don't do retries, etc.

File: curator-framework/src/main/java/org/apache/curator/framework/imps/RemoveWatchesBuilderImpl.java (hunk 12 above)
Comment (Randgalt) [inline curator-framework/src/main/java/org/apache/curator/framework/imps/RemoveWatchesBuilderImpl.java]: I was wrong. Ignore this.

File: curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesBuilder.java (hunk 7 above)
Comment (cammckenzie) [inline curator-framework/src/main/java/org/apache/curator/framework/api/RemoveWatchesBuilder.java]: Looking at the ZK code I think that 'local' mode occurs whenever there's any sort of error.

File: curator-framework/src/main/java/org/apache/curator/framework/imps/RemoveWatchesBuilderImpl.java (hunk 13 above)
Comment (Randgalt) [inline curator-framework/src/main/java/org/apache/curator/framework/imps/RemoveWatchesBuilderImpl.java]: Good point. You're right.

**Comment by cammckenzie** [comment]
All functionality is now implemented. Ready for a final review.

File: curator-framework/src/main/java/org/apache/curator/framework/api/DeleteBuilder.java [hunk 15]
```diff
@@ -18,6 +18,6 @@
  */
 package org.apache.curator.framework.api;
 
-public interface DeleteBuilder extends Guaranteeable, ChildrenDeletable
+public interface DeleteBuilder extends GuaranteeableDelete, ChildrenDeletable
```
Comment (Randgalt) [inline curator-framework/src/main/java/org/apache/curator/framework/api/DeleteBuilder.java]: A nit, but the naming convention is actually: <Verb1>..<VerbN>able. So it would be GuaranteeDeletable.

File: curator-framework/src/main/java/org/apache/curator/framework/api/BackgroundPathableQuietly.java [hunk 16]
```diff
@@ -0,0 +1,5 @@
+package org.apache.curator.framework.api;
+
+public interface BackgroundPathableQuietly<T> extends BackgroundPathable<T>, Quietly<BackgroundPathable<T>>
```
Comment (Randgalt) [inline curator-framework/src/main/java/org/apache/curator/framework/api/BackgroundPathableQuietly.java]: Naming convention nit: should be "BackgroundPathQuietlyable" - Yes, I'm probably going insane.

File: curator-framework/src/main/java/org/apache/curator/framework/imps/CuratorFrameworkImpl.java [hunk 17]
```diff
@@ -807,7 +821,7 @@ private void performBackgroundOperation(OperationAndData<?> operationAndData)
     {
         try
         {
-            if ( client.isConnected() )
+            if ( !operationAndData.isConnectionRequired() || client.isConnected() )
```
Comment (Randgalt) [inline curator-framework/src/main/java/org/apache/curator/framework/imps/CuratorFrameworkImpl.java]: Why this change?

File: curator-framework/src/main/java/org/apache/curator/framework/imps/FailedDeleteManager.java [hunk 18]
```diff
@@ -19,45 +19,18 @@
...
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
 
-class FailedDeleteManager
+class FailedDeleteManager extends FailedOperationManager<String>
 {
-    private final Logger log = LoggerFactory.getLogger(getClass());
-    private final CuratorFramework client;
-    
-    volatile FailedDeleteManagerListener debugListener = null;
```
Comment (Randgalt) [inline curator-framework/src/main/java/org/apache/curator/framework/imps/FailedDeleteManager.java]: nit: mark @VisibleForTesting

File: curator-framework/src/main/java/org/apache/curator/framework/imps/FailedRemoveWatchManager.java [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+    FailedRemoveWatchManager(CuratorFramework client)
+    {
+        super(client);
+    }
+
+    @Override
+    protected void executeGuaranteedOperationInBackground(FailedRemoveWatchDetails details)
+            throws Exception
+    {
+        if(details.watcher ==  null)
```
Comment (Randgalt) [inline curator-framework/src/main/java/org/apache/curator/framework/imps/FailedRemoveWatchManager.java]: nit: formatting

File: curator-framework/src/main/java/org/apache/curator/framework/imps/NamespaceWatcherMap.java [hunk 20]
```diff
@@ -70,6 +70,14 @@ NamespaceWatcher remove(Object key)
     {
         return map.remove(key);
     }
+    
+    /**
+     * Remove all watchers for a given path
+     * @param path
+     */
+    void removeAllForPath(String path) {
+
```
Comment (Randgalt) [inline curator-framework/src/main/java/org/apache/curator/framework/imps/NamespaceWatcherMap.java]: Why is this empty?

File: curator-recipes/src/test/java/org/apache/curator/framework/recipes/leader/ChaosMonkeyCnxnFactory.java [hunk 21]
```diff
@@ -59,13 +59,6 @@ public void startup(ZooKeeperServer zks) throws IOException, InterruptedExceptio
         super.startup(new ChaosMonkeyZookeeperServer(zks));
     }
 
-    /**
```
Comment (Randgalt) [inline curator-recipes/src/test/java/org/apache/curator/framework/recipes/leader/ChaosMonkeyCnxnFactory.java]: Why is this removed?

File: curator-framework/src/main/java/org/apache/curator/framework/imps/NamespaceWatcherMap.java (hunk 20 above)
Comment (cammckenzie) [inline curator-framework/src/main/java/org/apache/curator/framework/imps/NamespaceWatcherMap.java]: Looks like I started doing something with the NamespaceWatcherMap and then realised that I shouldn't. Will remove this method, it's not used anywhere.

File: curator-recipes/src/test/java/org/apache/curator/framework/recipes/leader/ChaosMonkeyCnxnFactory.java (hunk 21 above)
Comment (cammckenzie) [inline curator-recipes/src/test/java/org/apache/curator/framework/recipes/leader/ChaosMonkeyCnxnFactory.java]: Not sure? I don't see this change in my revision history? Last update was 6 months ago in the upgrade to ZK 3.5.0-alpha by Ioannis?

File: curator-framework/src/main/java/org/apache/curator/framework/api/DeleteBuilder.java (hunk 15 above)
Comment (cammckenzie) [inline curator-framework/src/main/java/org/apache/curator/framework/api/DeleteBuilder.java]: Ok, changed.

File: curator-framework/src/main/java/org/apache/curator/framework/api/BackgroundPathableQuietly.java (hunk 16 above)
Comment (cammckenzie) [inline curator-framework/src/main/java/org/apache/curator/framework/api/BackgroundPathableQuietly.java]: Changed. I have never written Quitelyable before!

File: curator-framework/src/main/java/org/apache/curator/framework/imps/FailedRemoveWatchManager.java (hunk 19 above)
Comment (cammckenzie) [inline curator-framework/src/main/java/org/apache/curator/framework/imps/FailedRemoveWatchManager.java]: Removed the extra space.

**Comment by Randgalt** [comment]
LGTM

**Comment by cammckenzie** [comment]
Ok, thanks will merge into the 3.0.0 branch.

**Comment by aaylward** [comment]
can this be closed?
